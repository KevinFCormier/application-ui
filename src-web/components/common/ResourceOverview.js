/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import React from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Loading } from 'carbon-components-react';
import { connect } from 'react-redux';
import CountsCardModule from '../CountsCardModule';
import ChannelsCardModule from '../ChannelsCardModule';
import StructuredListModule from '../../components/common/StructuredListModule';
import {
  getSingleResourceItem,
  resourceItemByName,
} from '../../reducers/common';
import {
  getNumDeployables,
  getNumDeployments,
  getNumFailedDeployments,
} from '../../../lib/client/resource-helper';
import { withLocale } from '../../providers/LocaleProvider';
import resources from '../../../lib/shared/resources';
import msgs from '../../../nls/platform.properties';

resources(() => {
  require('../../../scss/resource-overview.scss');
});

const ResourceOverview = withLocale(({ staticResourceData, item, params, modules, resourceType, showAppDetails, locale }) => {
  if (!item) { return <Loading withOverlay={false} className="content-spinner" />; }
  const modulesRight = [];
  const modulesBottom = [];
  React.Children.map(modules, (module) => {
    if (module.props.right) {
      modulesRight.push(React.cloneElement(module, {
        staticResourceData,
        resourceType,
        resourceData: item,
        params,
      }));
    } else {
      modulesBottom.push(React.cloneElement(module, {
        staticResourceData,
        resourceType,
        resourceData: item,
        params,
      }));
    }
  });

  const countsCardData = [
    {
      msgKey: 'table.header.deployables',
      count: getNumDeployables(item),
    },
    {
      msgKey: 'table.header.deployments',
      count: getNumDeployments(item),
    },
    {
      msgKey: 'table.header.failedDeployments',
      count: getNumFailedDeployments(item),
    },
  ];

  const channelsCardData = [
    {
      name: 'Development',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
    {
      name: 'QA',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
    {
      name: 'Dev',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
  ];

  return (
    <div className="overview-content">
      {!showAppDetails &&
        <React.Fragment>
          <div className="overview-content-bottom overview-content-with-padding">
            <CountsCardModule data={countsCardData} />
          </div>
          <div className="deployment-channels-title">
            {msgs.get('application.deployments.channels', locale)}
            {Array.isArray(channelsCardData) &&
              <span>&nbsp;({channelsCardData.length})</span>
            }
          </div>
          <div className="overview-content-bottom">
            <ChannelsCardModule data={channelsCardData} />
          </div>
        </React.Fragment>
      }
      {showAppDetails &&
        <StructuredListModule
          title={staticResourceData.detailKeys.title}
          headerRows={staticResourceData.detailKeys.headerRows}
          rows={staticResourceData.detailKeys.rows}
          data={item}
        />
      }
      {showAppDetails && modulesRight.length > 0 && (
        <div className="overview-content-right">{modulesRight}</div>
      )}
      {showAppDetails &&
        <div className="overview-content-bottom">{modulesBottom}</div>
      }
    </div>
  );
});

ResourceOverview.contextTypes = {
  locale: PropTypes.string,
};

ResourceOverview.propTypes = {
  item: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  modules: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  params: PropTypes.object,
  resourceType: PropTypes.object,
  staticResourceData: PropTypes.object,
};

const mapStateToProps = (state, ownProps) => {
  const { resourceType, params } = ownProps;
  const name = decodeURIComponent(params.name);
  const item = getSingleResourceItem(state, {
    storeRoot: resourceType.list,
    resourceType,
    name,
    predicate: resourceItemByName,
    namespace: params.namespace ? decodeURIComponent(params.namespace) : null,
  });
  return { item };
};

export default withRouter(connect(mapStateToProps)(withLocale(ResourceOverview)));
