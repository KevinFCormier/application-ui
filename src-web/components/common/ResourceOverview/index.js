/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { Loading } from 'carbon-components-react'
import { connect } from 'react-redux'
import CountsCardModule from '../../CountsCardModule'
import ChannelsCardCarousel from '../../ChannelsCardCarousel'
import ApplicationTopologyModule from '../../ApplicationTopologyModule'
import StructuredListModule from '../../../components/common/StructuredListModule'
import {
  getSingleResourceItem,
  resourceItemByName
} from '../../../reducers/common'
import {
  getChannelsList,
  getNumDeployables,
  getNumDeployments,
  getNumPendingDeployments,
  getNumInProgressDeployments,
  getNumFailedDeployments
} from './utils'
import { withLocale } from '../../../providers/LocaleProvider'
import resources from '../../../../lib/shared/resources'
import InstancesTable from '../../InstancesTable'

resources(() => {
  require('./style.scss')
})

const ResourceOverview = withLocale(
  ({
    staticResourceData,
    item,
    channelList,
    params,
    modules,
    resourceType,
    actions,
    showAppDetails,
    showExpandedTopology,
    getVisibleResources,
    incidentCount
  }) => {
    if (!item) {
      return <Loading withOverlay={false} className="content-spinner" />
    }
    const modulesRight = []
    const modulesBottom = []
    React.Children.map(modules, module => {
      if (module.props.right) {
        modulesRight.push(
          React.cloneElement(module, {
            staticResourceData,
            resourceType,
            resourceData: item,
            params
          })
        )
      } else {
        modulesBottom.push(
          React.cloneElement(module, {
            staticResourceData,
            resourceType,
            resourceData: item,
            params
          })
        )
      }
    })
    const countsCardData = [
      {
        msgKey: 'dashboard.card.deployables',
        textKey: 'dashboard.card.perInstance',
        count: getNumDeployables(item),
        border: 'right'
      },
      {
        msgKey: 'dashboard.card.deployments',
        textKey: 'dashboard.card.total',
        count: getNumDeployments(item)
      },
      {
        msgKey: 'dashboard.card.pending',
        textKey: 'dashboard.card.deployments',
        count: getNumPendingDeployments(item)
      },
      {
        msgKey: 'dashboard.card.inProgress',
        textKey: 'dashboard.card.deployments',
        count: getNumInProgressDeployments(item)
      },
      {
        msgKey: 'dashboard.card.failed',
        textKey: 'dashboard.card.deployments',
        count: getNumFailedDeployments(item),
        alert: true
      },
      {
        msgKey: 'dashboard.card.incidents',
        textKey: 'dashboard.card.total',
        count: incidentCount,
        border: 'left',
        alert: true
      }
    ]

    return (
      <div id="resource-overview" className="overview-content">
        {showAppDetails ? (
          <React.Fragment>
            <StructuredListModule
              title={staticResourceData.detailKeys.title}
              headerRows={staticResourceData.detailKeys.headerRows}
              rows={staticResourceData.detailKeys.rows}
              data={item}
            />
            {modulesRight.length > 0 && (
              <div className="overview-content-right">{modulesRight}</div>
            )}
            <div className="overview-content-bottom">{modulesBottom}</div>
          </React.Fragment>
        ) : !showExpandedTopology ? (
          <React.Fragment>
            <div className="overview-content-bottom overview-content-with-padding">
              <CountsCardModule data={countsCardData} />
            </div>
            {/* <div className="overview-content-bottom overview-content-with-padding">
              <ApplicationTopologyModule
                showExpandedTopology={showExpandedTopology}
                params={params}
                actions={actions}
              />
            </div> */}
            <div className="overview-content-bottom overview-content-with-padding-except-left">
              <ChannelsCardCarousel data={channelList} />
            </div>
            <div className="overview-content-bottom overview-content-with-padding">
              <InstancesTable
                staticResourceData={staticResourceData}
                resourceType={resourceType}
                getVisibleResources={getVisibleResources}
              />
            </div>
          </React.Fragment>
        ) : (
          <div className="overview-content-bottom overview-content-with-padding">
            <ApplicationTopologyModule
              showExpandedTopology={showExpandedTopology}
              params={params}
              actions={actions}
            />
          </div>
        )}
      </div>
    )
  }
)

ResourceOverview.contextTypes = {
  locale: PropTypes.string
}

ResourceOverview.propTypes = {
  item: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  modules: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  params: PropTypes.object,
  resourceType: PropTypes.object,
  staticResourceData: PropTypes.object
}

const mapStateToProps = (state, ownProps) => {
  const { resourceType, params } = ownProps
  const { HCMChannelList } = state
  const name = decodeURIComponent(params.name)
  const item = getSingleResourceItem(state, {
    storeRoot: resourceType.list,
    resourceType,
    name,
    predicate: resourceItemByName,
    namespace: params.namespace ? decodeURIComponent(params.namespace) : null
  })
  return { item, channelList: getChannelsList(HCMChannelList) }
}

export default withRouter(
  connect(mapStateToProps)(withLocale(ResourceOverview))
)
