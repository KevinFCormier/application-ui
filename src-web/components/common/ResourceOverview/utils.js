/*******************************************************************************
 * Licensed Materials - Property of IBM
 * 5737-E67
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
import R from 'ramda'
import {
  kindsToExcludeForDeployments,
  getResourcesStatusPerChannel
} from '../../ApplicationDeploymentPipeline/components/PipelineGrid/utils'
import { pullOutKindPerApplication } from '../../ApplicationDeploymentPipeline/utils'

// Method will take in an object and return back the channels mapped for display purposes
export const getChannelsList = channels => {
  if (channels && channels instanceof Array && channels.length > 0) {
    const mappedChannels = channels.map(channel => {
      // Will return back status as:
      // [0, 0, 0, 0, 0]
      // Pass, Fail, InProgress, Pending, Unidentified
      const status = getResourcesStatusPerChannel(channel, false)
      return {
        name: channel.name || '',
        counts: {
          completed: {
            total: status[0] + status[4]
          },
          inProgress: {
            total: status[2] + status[3]
          },
          failed: {
            total: status[1]
          }
        }
      }
    })
    return mappedChannels
  }
  return []
}

export const getNumClusters = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(elem => elem.kind === 'cluster')
    const numClusters = filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
    // number of clusters - 1 for local-cluster
    return numClusters > 0 ? numClusters - 1 : 0
  } else {
    return 0
  }
}

export const getNumDeployables = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(elem => elem.kind === 'deployable')
    return filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
  } else {
    return 0
  }
}

export const getNumDeployments = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(
      elem => !kindsToExcludeForDeployments.includes(elem.kind)
    )
    return filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
  } else {
    return 0
  }
}

export const getNumInProgressDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[2] + status[3]
}

export const getNumFailedDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[1]
}

export const getNumCompletedDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[0] + status[4]
}

export const getNumPolicyViolations = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    return 0
  } else {
    return 0
  }
}

// Given a current resource data INCLUDING its related resources
// we want to return all the channels.
// Channels are not returned inside related resources so we have to
// insepect each related subscription because it contains the channel
// -------------------------------------
// This method is not actually used but its got some good logic we could use later
// -------------------------------------
export const getResourceChannels = resourceData => {
  if (resourceData && resourceData.related) {
    const relatedData = resourceData.related
    // We want to pull only subscription data
    const filterToSubscriptions = x => x.kind == 'subscription'
    const subscriptionData = R.filter(filterToSubscriptions, relatedData)
    if (subscriptionData[0] && subscriptionData[0].items) {
      const channels = subscriptionData[0].items.map(sub => {
        return sub.channel
      })
      return channels
    }
    return []
  }
  return []
}

// Given an resource Object (item) and a list of resources Objects (appRelations)
// We want to go through the appRelations to see there is a resource UID that
// matches the item resource.
// If there is a match we want to return it and count it as common because
// it appears in both.
const findMatchingResource = (kind, item, appRelations) => {
  // Go through the relationship list
  const match = appRelations.map(resource => {
    const kindItems = resource.items || []
    const kindTwo = resource.kind || ''
    // go through each kinds items
    const matched = kindItems.map(itemTwo => {
      // If the UIDs match then we found a match
      if (kindTwo == kind && item._uid == itemTwo._uid) {
        return { kind: kind, items: [itemTwo] }
      }
    })
    // Filter out the undefined entries
    const filterOutUndefined = x => x != undefined
    const filteredData = R.filter(filterOutUndefined, matched)
    return filteredData
  })

  const filterOutEmpty = x => x.length > 0
  const filteredDataTwo = R.filter(filterOutEmpty, match)
  return filteredDataTwo
}

// We want to find the commone resources that between too objects
// EXAMPLE:
//  - We have one application with all of its related resources.
//  - We have a list of channels with all their relate resources.
//  - Given those TWO lists we want to find the common resources that are in
//    both the application related resources and the channels related resources.
//
// The reason is the channel resources wiill include ALL related resources in
// that particular channel which includes many different application resources
// if sepearate applications chose to put a subscription under that channel.
//
// -------------------------------------
// This method is not actually used but its got some good logic we could use later time
// -------------------------------------
export const getCommonResources = (channelData, appRelations) => {
  let channelFormat = []
  let commonResources = []
  // Go through Each Channel
  channelData.map(channel => {
    // That channels related data
    channel.related.map(resource => {
      const kindItems = resource.items || []
      const kind = resource.kind || ''
      // go through each kinds items
      kindItems.map(item => {
        const match = findMatchingResource(kind, item, appRelations)
        commonResources = commonResources.concat(match)
      })
    })
    // we want to tall up all the related resources for the given channel
    channelFormat = channelFormat.concat([
      { name: channel.name, items: commonResources }
    ])
  })
  // Filter out entries that are empty
  const filterOutUndefinedName = x => x.name != 'undefined'
  const filteredData = R.filter(filterOutUndefinedName, channelFormat)
  return filteredData
}

// We want to get the current Application if there is only ONE application.
// We check the secondaryHeader to see if there are two breadcrumbs which would
// mean that the user has selected one application to look into.
// We then extract the items from the application list and then pull out the application
// at location 0 which we know because there is only 1 application selected
// at this time.
export const getCurrentApplication = (HCMApplicationList, singleAppView) => {
  const currentApplication = R.pathOr([], ['items'])(HCMApplicationList)
  const currentApplicationItemZero =
    (singleAppView &&
      (currentApplication.length > 0 && currentApplication[0])) ||
    {}
  return currentApplicationItemZero
}

export const formatToChannel = (subscriptionList, bulkSubscription) => {
  let channelList = []
  subscriptionList instanceof Array &&
    subscriptionList[0] &&
    subscriptionList[0].items instanceof Array &&
    subscriptionList[0].items.map(subscription => {
      bulkSubscription &&
        bulkSubscription.items instanceof Array &&
        bulkSubscription.items.map(bulkSub => {
          const bulkSubChannelObj = pullOutKindPerApplication(
            bulkSub,
            'channel'
          )
          let bulkSubChannel = {}
          if (
            bulkSubChannelObj instanceof Array &&
            bulkSubChannelObj[0] &&
            bulkSubChannelObj[0].items instanceof Array &&
            bulkSubChannelObj[0].items[0]
          ) {
            bulkSubChannel = bulkSubChannelObj[0].items[0]
          }
          if (
            subscription.name == bulkSub.name &&
            subscription.namespace == bulkSub.namespace &&
            subscription.channel ==
              `${bulkSubChannel.namespace}/${bulkSubChannel.name}`
          ) {
            channelList = channelList.concat([
              {
                name: subscription.channel.split('/')[1],
                related: bulkSub.related
              }
            ])
          }
        })
    })
  return channelList
}

export const getSearchLinkForOneApplication = params => {
  if (params && params.name) {
    return `/multicloud/search?filters={"textsearch":"kind%3Aapplication%20name%3A${
      params.name
    }"}`
  }
  return ''
}

export const getSearchLinkForAllApplications = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Aapplication"}'
}

export const getSearchLinkForAllSubscriptions = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Asubscription"}'
}

export const getSearchLinkForAllChannels = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Achannel"}'
}
