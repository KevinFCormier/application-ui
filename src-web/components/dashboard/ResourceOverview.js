/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import lodash from 'lodash'
import PropTypes from 'prop-types'
import DashboardSection from '../common/DashboardSection'
import msgs from '../../../nls/platform.properties'
import { DashboardCard } from '../ComponentCard'
import {withRouter} from 'react-router-dom'

class ResourceOverview extends React.Component {
  render() {
    const { locale } = this.context
    const { cardItems = [] } = this.props
    return (
      <div id='component-summary'>
        <DashboardSection name={msgs.get('dashboard.section.resource-overview', locale)}>
          {cardItems.map(item =>
            item.name &&
            <DashboardCard
              critical={item.critical}
              warning={item.warning}
              healthy={item.healthy}
              type={item.type}
              title={msgs.get(`dashboard.card.${lodash.camelCase(item.name)}`, locale)}
              key={item.name}
              id={`card-${item.name.replace(' ', '-')}`}
            />)}
        </DashboardSection>
      </div>
    )
  }
}

DashboardSection.contextTypes = {
  locale: PropTypes.string
}

ResourceOverview.propTypes = {
  cardItems: PropTypes.arrayOf(PropTypes.shape({
    critical: PropTypes.number,
    healthy: PropTypes.number,
    warning: PropTypes.number,
    name: PropTypes.string,
    title: PropTypes.string,
    type: PropTypes.string,
  }))
}

export default withRouter(ResourceOverview)
