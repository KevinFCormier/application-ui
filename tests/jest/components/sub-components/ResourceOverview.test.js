/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router'
import ResourceOverview from '../../../../src-web/components/dashboard/ResourceOverview'

describe('ResourceOverview no components', () => {
  it('renders as expected', () => {
    const component = renderer.create(
      <MemoryRouter initialEntries={[ '/random' ]}>
        <ResourceOverview />
      </MemoryRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

const mockItem = {
  critical:1,
  warning:2,
  healthy:3,
  name:'4',
  table:[[1,2],[3,4]]
}

describe('ResourceOverview no components', () => {
  it('renders as expected', () => {
    const component = renderer.create(
      <MemoryRouter initialEntries={[ '/random' ]}>
        <ResourceOverview cardItem={mockItem} />
      </MemoryRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
