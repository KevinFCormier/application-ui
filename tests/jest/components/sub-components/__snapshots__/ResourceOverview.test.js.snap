// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceOverview no components renders as expected 1`] = `
<div
  id="resource-overview"
>
  <div>
    <div
      className="dashboard-section-modules"
    />
  </div>
</div>
`;

exports[`ResourceOverview no components renders as expected 2`] = `
<div
  id="resource-overview"
>
  <div>
    <div
      className="dashboard-section-modules"
    />
  </div>
</div>
`;
